File: app/services/ImdbApiService.java
package services;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;

import models.ImdbApiMovie;
import play.Logger;
import play.libs.WS;

import com.google.gson.Gson;
import com.google.gson.JsonElement;
import com.google.gson.JsonParser;

public class ImdbApiService {
	
	public ImdbApiMovie findOrRead(final String title){
		return findOrRead(title, null);
	}

	public ImdbApiMovie findOrRead(final String title, final Integer year){
		if(title==null){
			throw new IllegalArgumentException("NULL title");
		}
		ImdbApiMovie imdbMovie = null;//ImdbApiMovie.findByTitleAndYear(title, year);
		if(imdbMovie == null){
			String url = apiUrl(title, year);
			Logger.info(url);
			try{
				JsonElement json = new JsonParser().parse(WS.url(url).get().get().getBody());
				Gson gson = new Gson();
				imdbMovie = gson.fromJson(json, ImdbApiMovie.class);
				//imdbMovie.save();
			}catch(Exception ex){
				Logger.error(ex.getMessage(), ex);
			}
		}
		return imdbMovie;
	}

	private String apiUrl(final String title, final Integer year) {
		String url = null;
		try {
			url = "http://www.imdbapi.com/?t=" + URLEncoder.encode(title, "UTF-8");
			if(year != null){
				url = url + "&y=" +year;
			}
		} catch (UnsupportedEncodingException e) {
			throw new RuntimeException(e);
		}
		return url;
	}
}

File: app/services/ImdbApiService.scala
package services

import play.api.libs.json.{JsError, JsSuccess, Json}
import play.api.libs.ws.WSAPI
import play.api.Logger
import models.helper.ImdbApiMovie

import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future

object ImdbProtocol{
  implicit val imdbApiMovieReads = Json.reads[ImdbApiMovie]
}

object ImdbApiService {

}

class ImdbApiService(ws: WSAPI){

  private val logger = Logger("application.imdb")


  def find(title: String, year: Option[Int] = None): Future[Option[ImdbApiMovie]] = {

    // TODO fallback to a search without year if no result was found for a search including the year

    //val url = "http://www.imdbapi.com/"
    val url = "http://www.omdbapi.com/"
    val request = ws.url(url).withQueryString("t" -> title)
    val requestExtended = year.map(year =>
      request.withQueryString("y" -> year.toString)
    ).getOrElse(request)

    
    logger.info("Fetching " + requestExtended)

    requestExtended.get().flatMap{ response =>
      val json = response.json
      val errorOpt = (json \ "Error").asOpt[String]
      errorOpt match {
        case None =>
          import ImdbProtocol._
          Json.fromJson[ImdbApiMovie](response.json) match {
            case JsSuccess(movie, _) => Future.successful(Option(movie))
            case JsError(errors) => Future.failed(new RuntimeException(errors.toString()))
          }
        case Some(error) =>
          logger.warn(s"IMDB api internal error for $title $year: $error")
          // TODO should we also fail the future here?
          Future.successful(None)
      }
    }
  }

}